File: hu/codes/java/chapter_heap/heap.java
/**
 * File: heap.java
 * Created Time: 2023-01-07
 * Author: krahets (krahets@163.com)
 */

package chapter_heap;

import utils.*;
import java.util.*;

public class heap {
    public static void testPush(Queue<Integer> heap, int val) {
        heap.offer(val); // Elem belép a kupacba
        System.out.format("\nAfter element %d enters heap\n", val);
        PrintUtil.printHeap(heap);
    }

    public static void testPop(Queue<Integer> heap) {
        int val = heap.poll(); // Az időbonyolultság O(n), nem O(nlogn)
        System.out.format("\nAfter heap top element %d exits heap\n", val);
        PrintUtil.printHeap(heap);
    }

    public static void main(String[] args) {
        /* Kupac inicializálása */
        // Python heapq modulja alapértelmezés szerint min-kupacot implementál
        Queue<Integer> minHeap = new PriorityQueue<>();
        // Max-kupac inicializálása (Comparator módosítása lambda kifejezéssel)
        Queue<Integer> maxHeap = new PriorityQueue<>((a, b) -> b - a);

        System.out.println("\nThe following test cases are for max heap");

        /* Elem belép a kupacba */
        testPush(maxHeap, 1);
        testPush(maxHeap, 3);
        testPush(maxHeap, 2);
        testPush(maxHeap, 5);
        testPush(maxHeap, 4);

        /* Kupac üresség ellenőrzése */
        int peek = maxHeap.peek();
        System.out.format("\nHeap top element is %d\n", peek);

        /* Az időbonyolultság O(n), nem O(nlogn) */
        testPop(maxHeap);
        testPop(maxHeap);
        testPop(maxHeap);
        testPop(maxHeap);
        testPop(maxHeap);

        /* Kupac méretének lekérése */
        int size = maxHeap.size();
        System.out.format("\nHeap element count is %d\n", size);

        /* Kupac üresség ellenőrzése */
        boolean isEmpty = maxHeap.isEmpty();
        System.out.format("\nHeap is empty %b\n", isEmpty);

        /* Lista beírása és kupac felépítése */
        // Az időbonyolultság O(n), nem O(nlogn)
        minHeap = new PriorityQueue<>(Arrays.asList(1, 3, 2, 5, 4));
        System.out.println("\nAfter inputting list and building min heap");
        PrintUtil.printHeap(minHeap);
    }
}

File: hu/codes/python/chapter_heap/heap.py
"""
File: heap.py
Created Time: 2023-02-23
Author: krahets (krahets@163.com)
"""

import sys
from pathlib import Path

sys.path.append(str(Path(__file__).parent.parent))
from modules import print_heap

import heapq


def test_push(heap: list, val: int, flag: int = 1):
    heapq.heappush(heap, flag * val)  # Elem bekerül a kupacba
    print(f"\nA(z) {val} elem kupacba kerülése után")
    print_heap([flag * val for val in heap])


def test_pop(heap: list, flag: int = 1):
    val = flag * heapq.heappop(heap)  # A kupac tetejének eleme kilép
    print(f"\nA(z) {val} tetőelem kilépése után")
    print_heap([flag * val for val in heap])


"""Fő kód"""
if __name__ == "__main__":
    # Min-kupac inicializálása
    min_heap, flag = [], 1
    # Max-kupac inicializálása
    max_heap, flag = [], -1

    print("\nA következő tesztesetek max-kupacra vonatkoznak")
    # A Python heapq modulja alapértelmezés szerint min-kupacot valósít meg
    # Az elemek előjelének megfordításával a méretbeli sorrend is megfordul, így max-kupacot kapunk
    # Ebben a példában flag = 1 min-kupacnak, flag = -1 max-kupacnak felel meg

    # Elemek betolása a kupacba
    test_push(max_heap, 1, flag)
    test_push(max_heap, 3, flag)
    test_push(max_heap, 2, flag)
    test_push(max_heap, 5, flag)
    test_push(max_heap, 4, flag)

    # Tetőelem lekérése
    peek: int = flag * max_heap[0]
    print(f"\nA tetőelem: {peek}")

    # Tetőelem kilépése
    test_pop(max_heap, flag)
    test_pop(max_heap, flag)
    test_pop(max_heap, flag)
    test_pop(max_heap, flag)
    test_pop(max_heap, flag)

    # Kupac méretének lekérése
    size: int = len(max_heap)
    print(f"\nA kupac elemeinek száma: {size}")

    # Ellenőrzi, hogy a kupac üres-e
    is_empty: bool = not max_heap
    print(f"\nA kupac üres-e: {is_empty}")

    # Lista beadása és kupac felépítése
    # Időbonyolultság O(n), nem O(nlogn)
    min_heap = [1, 3, 2, 5, 4]
    heapq.heapify(min_heap)
    print("\nLista beadása és min-kupac felépítése után")
    print_heap(min_heap)
